File: 1026. Maximum Difference Between Node and Ancestor/maxAncestorDiff.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    void whoIsMyParent(TreeNode *node, vector<int> &papa, unordered_map<int, vector<int>> &m, int parent)
    {

        if (node == NULL)
            return;
        if (parent != -1)
        {
            papa.push_back(parent);
            m[node->val] = papa; // storing all the parents acc to that the child
        }
        whoIsMyParent(node->left, papa, m, node->val);
        whoIsMyParent(node->right, papa, m, node->val);

        papa.pop_back();
    }
    int maxAncestorDiff(TreeNode *root)
    {
        vector<int> papa; // papa will store the parents
        unordered_map<int, vector<int>> m;
        whoIsMyParent(root, papa, m, -1);

        int ans = 0;
        for (auto it : m)
        {

            for (auto val : it.second)
            {
                ans = max(ans, abs(it.first - val));
            }
        }
        return ans;
    }
};

File: 1026. Maximum Difference Between Node and Ancestor/maxAncestorDiff.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
public class Solution {
    public void whoIsMyParent(TreeNode node, List<Integer> papa, Map<Integer, List<Integer>> m, int parent) {
        if (node == null)
            return;

        if (parent != -1) {
            papa.add(parent);
            m.put(node.val, new ArrayList<>(papa)); // storing all the parents according to the child
        }

        whoIsMyParent(node.left, papa, m, node.val);
        whoIsMyParent(node.right, papa, m, node.val);

        if(parent!=-1)papa.remove(papa.size() - 1);
    }

    public int maxAncestorDiff(TreeNode root) {
        List<Integer> papa = new ArrayList<>(); // papa will store the parents
        Map<Integer, List<Integer>> m = new HashMap<>();
        whoIsMyParent(root, papa, m, -1);

        int ans = 0;
        for (Map.Entry<Integer, List<Integer>> entry : m.entrySet()) {
            int key = entry.getKey();
            List<Integer> values = entry.getValue();

            for (int val : values) {
                ans = Math.max(ans, Math.abs(key - val));
            }
        }
        return ans;
    }
}